cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc

```diff
@@ -172,6 +194,10 @@ auto non_zoned = timestamp(TimeUnit::NANO);
 #define DECLARE_TEMPORAL_BENCHMARKS_ZONED(OP) \
   BENCHMARK_TEMPLATE(BenchmarkTemporal, OP, zoned)->Apply(SetArgs);
 
+#define DECLARE_TEMPORAL_BINARY_BENCHMARKS(OP)                                \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalBinary, OP, non_zoned)->Apply(SetArgs); \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalBinary, OP, zoned)->Apply(SetArgs);
```

Comment:
For some cases you might want other time types here (date32 date64, time32, time64, duration).

Reply:
Hi @rok, thanks for the timely review! 

I was a bit curious about this, since it seems for `BenchmarkTemporalBinary` and `BenchmarkTemporal` (the original unary version), we are picking random `int64_t` -- which I assume is `date64`. How do we achieve the other time types, and is there some helpful example code I can follow?

Reply:
[A more complete example.](https://github.com/apache/arrow/blob/master/cpp/src/arrow/testing/generator.cc#L101-L155)

Reply:
Hi @rok, thank you for the clarification!

I refactored the code to work for the other data types, however it seems like these temporal binary functions [currently do not support any other datatype](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_temporal_binary.cc#L461) than `Int64`.

When running my code with something like `date32` or `time64` with `random.ArrayOf`, I get an `NotImplemented: Function years_between has no kernel matching input types (array[time64[ns]], array[time64[ns]])`

Reply:
Not sure why, but it seems like `date64` is viable as an input type, but `time64` / `time32`, and `duration` are not.

Reply:
Yeah, not all kernels will support all types.